### PR TITLE
Reduce Function Declarations

### PR DESCRIPTION
**Pull request by mitchute**
Pull request overview
---------------------
 - This PR reduces the functions declared in the header files to only those used by E+ outside of the associated cc file or those used in the unit tests.
 - I only checked on standard function return types, i.e. ```void```, ```int```, ```bool```, ```double```, ```Real64```, ```std::string```, and ```std::vector```.
 - cc files were reordered where necessary to preserve forward declaration.
 - Of the nearly 3000 functions in headers with those return types, this eliminates about 1200.
 - No appreciable compile-time differences were noted.

This is obviously a lot of changes, so if we need to break it down into smaller pieces to minimize merge conflicts, I'm happy to do that.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by amirroth** [comment]
I don't think that processing function declarations takes very long. You are just parsing and creating entries in a lookup table, not actually generating or optimizing code. I don't know that you would have seen much reduction in build time even if we weren't using pre-compiled headers ... which we are, yes?

**Comment by mitchute** [comment]
> I don't know that you would have seen much reduction in build time even if we weren't using pre-compiled headers ... which we are, yes?

We are using PCH I believe, but I know next to nothing about it. See here: https://github.com/NREL/EnergyPlus/blob/2ad67318cfe34c038d73081846e141c43440678b/CMakeLists.txt#L29-L42

That snippet sort of seems like we are only doing a subset, but again, I haven't dug in enough to understand it.

**Comment by amirroth** [comment]
PCH are the cached CTFs of compilers. That analogy actually works much better than I thought at first.

**Comment by mbadams5** [comment]
How does this affect IDEs such as Visual Studio, CLion, etc. when doing code completion? Does it do code completion of the functions only within the same CC file? Will this reduce function discoverability since you'll have to go digging through CC files to find a function to potentially use that is not currently shared across CC files?
This also means we have some order dependence in the CC files with regard to functions where we didn't previously.

Also it seems like we'll be reverting parts of the work as we need functions for future unit tests or future cross CC file uses?

I guess I don't understand the point of this effort, especially if it doesn't improve compilation speed (which like Amir stated is unlikely for the function declarations in the headers).

Are we trying to limit the "public" API of functions? But we already do that for the Python API I thought.

It just seems like a lot of unnecessary changes to me, but I may be missing the point or benefit.

**Comment by Myoldmopar** [comment]
I've never seen a project that they just added every one of their functions to the header files.  I would say it is a bad habit to expose all functions through the header files just because that's how it was translated from Fortran.  

As for reverting the parts of the work later, for one thing that will be a minor 'revert' to simply add the function to the header file, but more importantly, that will help the developer understand that this function was not previously intended to be called from outside that file, and they are now exposing it.  

As for IDEs, yes, it definitely now only limits to what the code can see based on current header declarations, which is a fantastic thing.  Headers should only expose the meaningful part of the code through the header so that a developer understands immediately that only select functions are set up to be called from outside.  

This argument could be applied to objects, right?  If you mark a method as private then the IDE won't pick it up....because it's designed that way.  And if you need to make a method public, then you should take pause and think about why you are wanting to call that method in a new way.

**Comment by mbadams5** [comment]
Maybe we are looking at different projects, but many I look at do include most or all functions in the header. NOW that said, they do not expose all of them to the public API. Usually they'll use different namespaces to differentiate between public/export and internal/private. Or they will use a public header and an internal header or use a PIMPL implementation. Plus with classes/structs you have to declare all functions in the header but yes, at least with those you can specify public/protected/private as part of the class/struct.

I mean most of this is due to everything being in the same global namespace (and yes, driven by Fortran and the conversion) and limitations of headers in C/C++.

If the point is to restrict E+ developers from using a function, why not use anonymous namespaces in the header since that doesn't allow the function to be used outside that module (translation unit)? If the point is to restrict the public API for external developers to use E+ as a library then maybe we should have public headers and internal headers or some other similar method?

I believe the best answer is to wait for C++20 modules ([cppreference](https://en.cppreference.com/w/cpp/language/modules) or [modules overview](https://docs.microsoft.com/en-us/cpp/cpp/modules-cpp?view=msvc-160)). This will allow you to mark specific functions for `export` then you `import` the module and can only use the exported functions. Modules do have the benefit of reducing compilation times.

I think introducing order dependence for some of the functions in the implementation files isn't ideal. Plus personally, I like being able to see all functions in a file in a single location where it is more succinct and includes a short description of the function and it's use.

On a final note, if we have 1200 of 3000 functions that are not used by other functions outside the current file nor in unit tests directly that means we should probably have better unit tests (or it maybe shouldn't be a standalone function to begin with).

**Comment by amirroth** [comment]
PIMPL! We used PIMPL at Microsoft, with I- and C- classes and everything. It was wild stuff. I don't think that approach will fly with the current anti-pointer sentiment in the project. PIMPL also adds runtime overhead which I don't think we want. I hadn't heard of anonymous `namespace`s or what the plans are for C++20 `namespace` features, but using those features to control function visibility sounds like a better approach. In either case, EnergyPlus definitely needs a better `namespace` regime. There is already [a little something](https://github.com/NREL/EnergyPlus/wiki/File-and-Module-Structure#namespaces-and-namespace-resolution) in the wiki about it, but maybe not enough.

If we are worried about the proliferation of functions in EnergyPlus (beyond simply having multiple copies of LUdecomposition), one approach is to increase the use of inheritance with more heavyweight parent classes. For instance, there is a `PlantComponent` class that all plant objects inherit from, but that class is relatively thin (it has only one data member five functions, all of which are either pure or degenerate) and it looks like the hierarchy is relatively flat, i.e., most objects inherit directly from `PlantComponent`. Some redundancy can be reduced by making `PlantComponent` a thicker object (with more data members and methods) and/or by developing intermediate `class`es that reduce duplication (of both functions and data) at leaf classes.

File: src/EnergyPlus/DXCoils.hh [hunk 1]
```diff
@@ -693,20 +676,8 @@ namespace DXCoils {
...
-    void CalcTwoSpeedDXCoilStandardRating(EnergyPlusData &state, int const DXCoilNum);
-
-    void GetFanIndexForTwoSpeedCoil(
-        EnergyPlusData &state, int const CoolingCoilIndex, int &SupplyFanIndex, std::string &SupplyFanName, int &SupplyFan_TypeNum);
-
-    Real64 CalcTwoSpeedDXCoilIEERResidual(EnergyPlusData &state,
-                                          Real64 const SupplyAirMassFlowRate, // compressor cycling ratio (1.0 is continuous, 0.0 is off)
-                                          std::array<Real64, 12> const &Par   // par(1) = DX coil number
-    );
-
```
Comment (jmythms) [inline src/EnergyPlus/DXCoils.hh:709]: Very cool!

**Comment by Myoldmopar** [comment]
> Maybe we are looking at different projects

I *did my own research* :grimacing: and opened up the kernel code :one:.  The very first function I looked at (`cpuhp_get_step`) is declared in the c file, used a ton of times within that c file, and not declared in a header.  [cpuhp_get_step](https://github.com/torvalds/linux/blob/7041503d3a5c869e4b4934df57112ef90ce7e307/kernel/cpu.c#L143-L146)

> why not use anonymous namespaces

This is fine, let's do that.  I added more discussion below after my venting :+1: 

> maybe we should have public headers

We already do break out [public API](https://github.com/NREL/EnergyPlus/tree/develop/src/EnergyPlus/api) headers separately, this is purely an internal unit-to-unit change. 

Frankly, I thought this would be a harmless little change.  We want to get deeper refactors of EnergyPlus, and it seemed like an easy way to start characterizing function usage, and making the function visibility more intentional was to clean out the unused ones from the headers.  That's literally the whole intent here.  

I saw an added benefit in an IDE, where you would only see which functions were purposely made available from the file because only those would live in the header.  Then you don't have to sift through the 25 internal worker functions listed for a namespace when the only **valid** entry point into that namespace is the `SimCoil` method or whatever.

If this causes too much heartburn then we'll just abandon it, it's not something I'm all that passionate about.  I really just thought it would be a positive, relatively easy, step forward in refactoring without causing any diffs or issues.

If we still want to do this, and anonymous namespaces are the way to go, that is fine, just let me know what should show up in the header.  In this example, the library code has a `calculate()` and a `simulate()` method.  The `simulate()` method is the only valid entry point, but currently `calculate()` also sits in the header, so a new developer could mistakenly call this and not understand why it won't work.  

```c++
//  in library.h
namespace EnergyPlus::Coil {
    namespace {
        // this anonymous namespace is essentially how we would 
        // label global hanging functions as "private:"
        int calculate();
    }
    int simulate();
}

//  in library.cpp
namespace EnergyPlus::Coil {
    int calculate() {
        return 42;
    }
    int simulate() {
        return calculate();
    }
}

//  in main.cpp
int main() {
    return EnergyPlus::Coil::simulate();
}
```

When I put the calculate method into an anonymous namespace in the header, the compiler doesn't like it, because now you have two possible distinct declarations of `calculate()`.  Are you instead suggesting moving the function to an anonymous namespace in the cpp file?  I think I'm just missing how we should do that, and I am all ears.

Thanks for the feedback all.

--------------------

:one: I am uninterested in expanded discussions comparing EnergyPlus and the kernel...

**Comment by Myoldmopar** [comment]
OK, with the constexpr changes merged in, this is now super conflicted.  @mitchute I'm curious what it would take to resolve the conflicts vs just start over.  Either way, as much as I want this to proceed, this is not a priority task, so I'm going to close the PR for now.  I'll leave the branch here so we could bring this back up soon.  The Elders of the Internet have spoken and think that global function declarations should not muddy up headers.  If people here still have thoughts on how we could do a compromise to clean up the headers in a better way, please comment.

**Comment by mitchute** [comment]
> I'm curious what it would take to resolve the conflicts vs just start over.

Resolve conflicts - 👎 
Start over - 👍

**Pull request closed without merging**